server/src/main/java/org/opensearch/action/admin/indices/segments/TransportPitSegmentsAction.java

```diff
@@ -0,0 +1,261 @@
+                            ShardRoutingState.STARTED,
+                            null,
+                            null,
+                            null,
+                            -1L
+                        )
+                    );
+                }
+            }
+        }
+        return new PlainShardsIterator(iterators);
+    }
+
+    @Override
+    protected ClusterBlockException checkGlobalBlock(ClusterState state, PitSegmentsRequest request) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected ClusterBlockException checkRequestBlock(ClusterState state, PitSegmentsRequest countRequest, String[] concreteIndices) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
+    }
+
+    @Override
+    protected ShardSegments readShardResult(StreamInput in) throws IOException {
+        return new ShardSegments(in);
+    }
+
+    @Override
+    protected IndicesSegmentResponse newResponse(
+        PitSegmentsRequest request,
+        int totalShards,
+        int successfulShards,
+        int failedShards,
+        List<ShardSegments> results,
+        List<DefaultShardOperationFailedException> shardFailures,
+        ClusterState clusterState
+    ) {
+        return new IndicesSegmentResponse(
+            results.toArray(new ShardSegments[results.size()]),
+            totalShards,
+            successfulShards,
+            failedShards,
+            shardFailures
+        );
+    }
+
+    @Override
+    protected PitSegmentsRequest readRequestFrom(StreamInput in) throws IOException {
+        return new PitSegmentsRequest(in);
+    }
+
+    @Override
+    public List<ShardRouting> getShardRoutingsFromInputStream(StreamInput in) throws IOException {
+        return in.readList(PitAwareShardRouting::new);
+    }
+
+    /**
+     * This retrieves segment details of PIT context
+     * @param request      the node-level request
+     * @param shardRouting the shard on which to execute the operation
+     */
+    @Override
+    protected ShardSegments shardOperation(PitSegmentsRequest request, ShardRouting shardRouting) {
+        if (shardRouting instanceof PitAwareShardRouting) {
+            PitAwareShardRouting pitAwareShardRouting = (PitAwareShardRouting) shardRouting;
+            SearchContextIdForNode searchContextIdForNode = decode(namedWriteableRegistry, pitAwareShardRouting.getPitId()).shards()
+                .get(shardRouting.shardId());
+            PitReaderContext pitReaderContext = searchService.getPitReaderContext(searchContextIdForNode.getSearchContextId());
+            if (pitReaderContext == null) {
+                return new ShardSegments(shardRouting, new ArrayList<>());
+            }
+            return new ShardSegments(pitReaderContext.getShardRouting(), pitReaderContext.getSegments());
+        }
+        return new ShardSegments(shardRouting, new ArrayList<>());
```

Comment:
In what scenario the control flow reaches this return statement?

Reply:
if in case shard routing is not instance of pit shard routing , then it'll reach here

Reply:
In what conditions `shardOperation()` be called with shardRouting that is not an instance of `PitAwareShardRouting`?
I want to understand if this is a valid scenario. If not, should we provide empty list or throw an exception?

Reply:
as per the flow , this should almost never happen as shards() method will only create pitawareshardrouting iterator and then the shardoperation operates on the same.

Reply:
Got it, then what is the expected behavior `if` control reaches here? Returning empty list vs throwing exception?

Reply:
Better to throw exception as this code should not be hit in happy/correct flow - will make changes for that. Wdyt ?